### DISCUSSION TITLE
How do you add another compiler to Ghidra

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#6203] [Q&A]
Currently Ghidra has several built-in compiler specific details (like those included in `windows_v12_32.gtd` etc.).  I'd like to create something similar for MSVC4.20.  I've discovered that those PE files have a major/minor linker attribute available in the Optional PE Header that would identify the compiler, but I've no clue as to how to go about integrating it!

**Comment by lightspot21** [comment]
What kind of details are you looking to add? As far as I understand, `.gdt` files store data type definitions, (usually) made by feeding C headers into Ghidra. Can you clarify the use case?

**Comment by Wall-AF** [comment]
Once you've created your `.gtd` file and know the magic linker numbers (that identify the compiler), the next time you import something how does Ghidra recogise your new `gtd` and compiler?

**Comment by lightspot21** [comment]
I think the keywords to look for are 'loader' and 'opinion'. Here's what I'd do if I were you:

1. Extend Ghidra's PE loader (see https://github.com/NationalSecurityAgency/ghidra/tree/master/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/PeLoader.java and the files around it) with any heuristics that identify the specific compiler version. 
2. Tell the loader to attach the proper gdt file during the `Apply Data Types` step. 
3. You should be able to automatically use your custom `gdt` file on any programs that are detected by the loader to be made with the specific compiler.

(I haven't ever had a need to do this, so take these with adequate amounts of salt. If anyone from the Ghidra team or a more experienced user is willing to correct me, please feel free to do so)

**Reply by Wall-AF** [reply]
Looks a bit nuanced!  Need more advice if poss!

**Reply by lightspot21** [reply]
I went to dig in the source a bit and apparently I was wrong. Here's what actually happens:

Ghidra decides which gdt to apply based on some information about the loaded program. The `Apply Data Archives` analyzer located at `Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/analysis/ApplyDataArchiveAnalyzer.java`, via its `getAutoDTMs()` method (which is the one called if autodetection is desired), calls out to `Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/datamgr/util/DataTypeArchiveUtility.java`, specifically its `getArchiveList()` method.

Now, `getArchiveList()` returns all applicable gdts, and it checks whether a gdt could be a match based on heuristics (e.g. if it's a 64-bit PE file, add `windows_vs12_64.gdt` to the list of matches). I guess to have Ghidra return your gdt file automatically, you should modify `getArchiveList()` in order to read off the linker attribute you discovered and based on that, return your own gdt file instead of the default.

**Reply by ryanmkurtz** [reply]
I think Ghidra will automatically discover your GDT file if it resides in some module's `data/` directory.  Ultimately though, it will be the job of your `Analyzer` to apply it.  You might find the golang example useful.  Here is the starting point: 

https://github.com/NationalSecurityAgency/ghidra/blob/d8dae6d0928f8aa68513b7a50265bf0cc63a5388/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/golang/rtti/GoRttiMapper.java#L218-L241
